subsys/net/lib/mqtt_sock/mqtt.c

```diff
@@ -627,3 +597,43 @@ int mqtt_input(struct mqtt_client *client)
 
 	return err_code;
 }
+
+int mqtt_read_publish_payload(struct mqtt_client *client, void *buffer,
+			      size_t length)
+{
+	int ret;
+
+	NULL_PARAM_CHECK(client);
+
+	mqtt_mutex_lock(client);
+
+	if (client->internal.remaining_payload == 0) {
+		ret = -EPERM;
```

Comment:
This is perfectly normal condition, EOF reached. Just return EOF (== length of 0).